### ISSUE TITLE
Possible bug with RSACng hash verification

### ISSUE THREAD

**Issue by EatonZ** [dotnet/corefx#36343]
Original question: https://stackoverflow.com/questions/55319441/rsacng-fails-to-verify-hash
I've verified the problem also affects .NET Core and am convinced this may be a bug in both .NET Core and .NET Framework.

**How to reproduce:**
1. Create a new .NET Core Console App in VS 2017 on Windows.
2. Add the `System.Security.Cryptography.Cng` package.
3. Add `using System.Security.Cryptography;` to the top of `Program.cs` and paste this code into `Main`:
```
var rsaKeyPub = "BgIAAACkAABSU0ExAAgAAAMAAAC7N02Zb0lf3UO4pl3ymFvkrNSkPP0Q076vYzvbeTqS5vkBcoXloM044q1LCFtaXw6DUzSFM0IqoGOONb+PW+UeNcTcA/+MKNi7nzbBGg3kAj8QhuxO3u2QJXg62Zb9H+SEvZYfi9PhOCSo0LpWKl72k+uaoTyPAuV738TamRvXWb8XOswSmsHQa38q1Id4TW7CzvVOjc0vnhL+rZ8Po1qg5FJc8m8gdGWC0a4NJTzBsOqLzeVE12B8zgIMehu1gGw/SjY5PVEkDABWgY2DzxLT3rbs6oZ5ZLSHu041q3s1ihOQ8+GMRx3qqvPyB4JVlyd7jqN0j0dT+Yqr+8t3/Liu";
var hash = "Ow1kg47GAgf9cyZbisDuTRNy5NQ=";
var signature = "bBnFouYvuZSYZIPihDB4J/CVC7o5ej3MrbkZV9cn6vgL23rDW1jevWEHx4wGBXLc443DKrZ0XQlSpp3FE/+isyDMcGh7c0buMufiYuOQ0rbo8e4tvuZuZpt+06xnBQcYyFMqe4lkFcI0f/NeAIvy1vME+Kq4v3ikwR4+CsjObgEJIBdWB0B4cqp8355pxtYJv2BQ7UHy/Tv0+OtslgbxikrwU2CQ+tR3XHywIdzm0BEOBfdnOlky96ED18BAqwLlxjef0snCl3DvKz93gtIIQVwEoDRlKC/v/Xb4Eke/fyvt66orLEIyL8Emaer9J6P38ZB1pWRuOsLCv4ly8fnOMw==";
RSAParameters rsaParams;
using (var rsa = new RSACryptoServiceProvider())
{
    rsa.ImportCspBlob(Convert.FromBase64String(rsaKeyPub));
    //Export now for easy importing in the next function. This is not the reason the next block returns false.
    rsaParams = rsa.ExportParameters(false);
    var valid = rsa.VerifyHash(Convert.FromBase64String(hash), Convert.FromBase64String(signature), HashAlgorithmName.SHA1, RSASignaturePadding.Pkcs1);
}
using (var rsa = new RSACng())
{
    rsa.ImportParameters(rsaParams);
    var valid = rsa.VerifyHash(Convert.FromBase64String(hash), Convert.FromBase64String(signature), HashAlgorithmName.SHA1, RSASignaturePadding.Pkcs1);
}
```
4. Set breakpoints after `var valid` to see the results. The first block returns true, the second false.

This bug may go pretty deep, all the way down into `NCryptVerifySignature`. I don't know where I would post a bug report for that, though. Maarten Bodewes on Stack Overflow suggests it may be a problem with the exponent (3) this particular key uses.

**Comment by vcsjones** [comment]
Similar behavior was reported in https://github.com/dotnet/corefx/issues/34202#issuecomment-450587790. Checking to see if it's the same issue.

@EatonZ can you share what software created the signature?

**Comment by EatonZ** [comment]
@vcsjones This particular signature was generated in a factory as part of a manufacturing process. I do not know the specifics of the tools used to generate it, but know it's verified with a "Pkcs1Verify" series of functions. Also, the signature needed to be reversed (using `Array.Reverse`) before it would verify with `RSACryptoServiceProvider`. Not reversing it doesn't change `RSACng`'s behavior.

**Comment by vcsjones** [comment]
So running it through OpenSSL via `openssl rsautl -verify -in sig.bin -inkey pub.key -pubin -hexdump -raw`

Gives back

<pre>
0000 - 00 01 ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
0010 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
0020 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
0030 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
0040 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
0050 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
0060 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
0070 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
0080 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
0090 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
00a0 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
00b0 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
00c0 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
00d0 - ff ff ff ff ff ff ff ff-ff ff ff ff ff ff ff ff   ................
00e0 - ff ff ff ff ff ff ff ff-ff ff ff 00 3b 0d 64 83   ............;.d.
00f0 - 8e c6 02 07 fd 73 26 5b-8a c0 ee 4d 13 72 e4 d4   .....s&[...M.r..
</pre>

This appears to be the same problem as https://github.com/dotnet/corefx/issues/34202#issuecomment-450589549

@EatonZ the issue boils down to the signature not having a fully complete PKCS#1.5 signature that includes the OID. CNG is pickier about the signature missing that OID, while CAPI allows for it. CNG _does_ support it, but you need to pass options that .NET Core nor Framework expose at the moment.

/cc @bartonjs

**Comment by EatonZ** [comment]
@vcsjones Interesting. So `RSACryptoServiceProvider` is less picky?
Is there going to be any work to fix this shortcoming? That comment thread reads to me like this won't be addressed any time soon.

**Comment by bartonjs** [comment]
@EatonZ No, there is no plan at this time to allow verifying those types of signatures.  They are not valid according to the public specifications.

**Comment by bartonjs** [comment]
Closing as (effectively) duplicate of #34202.

**Comment by EatonZ** [comment]
@karelz I noticed you've added this to the 3.0 milestone. Does that mean plans have changed and this will be addressed by 3.0's release?

**Comment by karelz** [comment]
The issue was closed during 3.0 milestone, that why it is milestoned as such (bulk edit). You should monitor the primary issue - which is without milestone at the moment.
